File: race3.py
"""
ID: totheso1
LANG: PYTHON3
TASK: race3
"""

from collections import deque
from typing import Any

fin = open("race3.in", "r")
fout = open("race3.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
adjacency_list = []


def initialize() -> None:
    for f, line in enumerate(fin):
        int_collection = list(map(int, line.split()))
        if int_collection[0] != -1:
            int_collection.pop()
            adjacency_list.append(int_collection)


def solve() -> None:
    def bfs_avoid(avoid: int) -> bool:
        que = deque((0,))
        first_bfs_visited[0] = True
        while que:
            curr_vertex = que.popleft()
            for to in adjacency_list[curr_vertex]:
                if first_bfs_visited[to] or to == avoid:
                    continue
                if to == vertex_count - 1:
                    return True
                que.append(to)
                first_bfs_visited[to] = True
        return False

    def bfs_can_go_back(start: int) -> bool:
        visited = [False] * vertex_count
        que = deque((start,))
        visited[start] = True
        while que:
            curr_vertex = que.popleft()
            for to in adjacency_list[curr_vertex]:
                if visited[to]:
                    continue
                if first_bfs_visited[to]:
                    return True
                que.append(to)
                visited[to] = True
        return False

    unavoidable_point_collection = []
    splitting_point_collection = []
    vertex_count = len(adjacency_list)
    for vertex in range(vertex_count - 1):
        first_bfs_visited = [False] * vertex_count
        if not bfs_avoid(vertex):
            unavoidable_point_collection.append(vertex)
            if not bfs_can_go_back(vertex):
                splitting_point_collection.append(vertex)
    fprint(len(unavoidable_point_collection), *unavoidable_point_collection)
    fprint(len(splitting_point_collection), *splitting_point_collection)


initialize()
solve()
# Code end

File: race3.cpp
/*
ID: totheso1
LANG: C++14
TASK: race3
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

static ifstream fin("race3.in");
static ofstream fout("race3.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static vector<vector<int>> adjacency_list(1);

static void initialize() {
  for (int input; fin >> input, input != -1;) {
    if (input == -2) {
      adjacency_list.emplace_back();
    } else {
      adjacency_list.back().push_back(input);
    }
  }
  adjacency_list.pop_back();
}

static void solve() {
  const int vertex_count = adjacency_list.size();
  vector<bool> first_bfs_visited;
  const auto bfs_avoid = [&](const int avoid) -> bool {
    queue<int> que({0});
    first_bfs_visited[0] = true;
    while (!que.empty()) {
      const int curr_vertex = que.front();
      que.pop();
      for (const int to : adjacency_list[curr_vertex]) {
        if (first_bfs_visited[to] || to == avoid) {
          continue;
        }
        if (to == vertex_count - 1) {
          return true;
        }
        que.push(to);
        first_bfs_visited[to] = true;
      }
    }
    return false;
  };

  const auto bfs_can_go_back = [&](const int start) -> bool {
    vector<bool> visited(vertex_count, false);
    queue<int> que({start});
    visited[start] = true;
    while (!que.empty()) {
      const int curr_vertex = que.front();
      que.pop();
      for (const int to : adjacency_list[curr_vertex]) {
        if (visited[to]) {
          continue;
        }
        if (first_bfs_visited[to]) {
          return true;
        }
        que.push(to);
        visited[to] = true;
      }
    }
    return false;
  };

  vector<int> unavoidable_point_collection;
  vector<int> splitting_point_collection;
  for (int vertex = 0; vertex < vertex_count - 1; ++vertex) {
    first_bfs_visited.assign(vertex_count, false);
    if (!bfs_avoid(vertex)) {
      unavoidable_point_collection.push_back(vertex);
      if (!bfs_can_go_back(vertex)) {
        splitting_point_collection.push_back(vertex);
      }
    }
  }
  fout << unavoidable_point_collection.size();
  for (const int v : unavoidable_point_collection) {
    fout << ' ' << v;
  }
  fout << endl << splitting_point_collection.size();
  for (const int v : splitting_point_collection) {
    fout << ' ' << v;
  }
  fout << endl;
}

int main() {
  initialize();
  solve();
}
